contracts/lib/ExecutionLib.sol

```diff
@@ -50,38 +54,44 @@ library ExecutionLib {
 
     /**
      * @notice Decode a batch of `Execution` pointers from a `bytes` calldata.
-     * @dev code is copied from solady's LibERC7579.sol
-     * https://github.com/Vectorized/solady/blob/ceb1019a774a04c2ee1a283d3db5a3dfcb7ab8c2/src/accounts/LibERC7579.sol#L149
-     *      Credits to Vectorized and the Solady Team
+     * https://github.com/OpenZeppelin/openzeppelin-contracts/pull/5353
      */
-    function decodeBatch(bytes calldata executionData) internal pure returns (Execution[] calldata pointers) {
-        /// @solidity memory-safe-assembly
-        assembly {
-            let u := calldataload(executionData.offset)
-            if or(shr(64, u), gt(0x20, executionData.length)) {
-                mstore(0x00, 0xba597e7e) // `DecodingError()`.
-                revert(0x1c, 0x04)
+    function decodeBatch(bytes calldata executionCalldata)
+        internal
+        pure
+        returns (Execution[] calldata executionBatch)
+    {
+        unchecked {
+            uint256 bufferLength = executionCalldata.length;
+
+            // Check executionCalldata is not empty.
+            if (bufferLength < 32) revert ERC7579DecodingError();
+
+            // Get the offset of the array (pointer to the array length).
+            uint256 arrayLengthPointer = uint256(bytes32(executionCalldata[0:32]));
+
+            // The array length (at arrayLengthPointer) should be 32 bytes long. We check that this is within the
+            // buffer bounds. Since we know bufferLength is at least 32, we can subtract with no overflow risk.
+            if (arrayLengthPointer > bufferLength - 32) revert ERC7579DecodingError();
+
+            // Get the array length. arrayLengthPointer + 32 is bounded by bufferLength so it does not overflow.
+            uint256 arrayLength = uint256(bytes32(executionCalldata[arrayLengthPointer:arrayLengthPointer + 32]));
+
+            // Check that the buffer is long enough to store the array elements as "offset pointer":
+            // - each element of the array is an "offset pointer" to the data.
+            // - each "offset pointer" (to an array element) takes 32 bytes.
+            // - validity of the calldata at that location is checked when the array element is accessed, so we only
+            //   need to check that the buffer is large enough to hold the pointers.
+            //
+            // Since we know bufferLength is at least arrayLengthPointer + 32, we can subtract with no overflow risk.
+            // Solidity limits length of such arrays to 2**64-1, this guarantees `arrayLength * 32` does not overflow.
+            if (arrayLength > type(uint64).max || bufferLength - arrayLengthPointer - 32 < arrayLength * 32) {
```

Comment:
I have a question about this condition `bufferLength - arrayLengthPointer - 32 < arrayLength * 32`
why `arrayLength` is multiplied by `32`?
The length of an element of an array is not 32. it is at least 32 (address to) + 32 (value) + 32 (calldata offset) + 32 (calldata length) = 128

```
Logs:
  arrayLength =  3
  arrayLength*32 =  96
  bufferLength =  704
  arrayLengthPointer =  32
  bufferLength - arrayLengthPointer - 32 =  640
  Length of one element =  224
  ```
  
  this is from one of the tests.
  
So even if `X = bufferLength - arrayLengthPointer - 32` is too small to fit all the array elements, `arrayLength * 32` can still be less than `X`

Reply:
>  - each element of the array is an "offset pointer" to the data.

Reply:
ah hell. makes sense. thanks for pointing this.